src/node/eviction.cpp

```diff
@@ -243,6 +244,53 @@ void ProtectEvictionCandidatesByRatio(std::vector<NodeEvictionCandidate>& evicti
 EvictionManagerImpl::EvictionManagerImpl() {}
 EvictionManagerImpl::~EvictionManagerImpl() = default;
 
+void EvictionManagerImpl::AddCandidate(NodeId id, std::chrono::seconds connected,
+                                       uint64_t keyed_net_group, bool prefer_evict,
+                                       bool is_local, Network network,
+                                       bool noban, ConnectionType conn_type)
+{
+    LOCK(m_candidates_mutex);
+    // The default values used to initialise the eviction candidates result in
+    // newer candidates being more likely to be evicted.
+    NodeEvictionCandidate candidate{
```

Comment:
I wonder if it might make sense to split NodeEvictionCandidate into 2 parts: the const properties that don't change (like conn_type/noban) and the non-const state that's intended to be/might be updated?

Maybe something like:

```c++
class NodeEvictionCandidate {
  class Properties {
    NodeId m_id;
    ConnectionType m_conn_type;
    bool m_noban;
    ...
  };
  NodeEvictionCandidate(NodeProperties properties) : m_properties(properties){}
private:
  const NodeProperties m_properties;
  bool m_relay_txs;
  bool fBloomFilter;
  ...
};
...
void EvictionManagerImpl::AddCandidate(NodeEvictionCandidate::Properties props) {...}
```

I think that might be more straightforward to use as it's self-documenting, while also allowing for safe lock-free access to the stuff that can't change.



Maybe that's easier said than done though. Thoughts?

Reply:
Even if you don't do this, I'd suggest marking the const members const, and group them before the mutable members.

Reply:
I tried to do this but the `src/test/net_peer_eviction_tests.cpp` tests sadly rely on no members of `NodeEvictionCandidate` being const. Would like to leave this for a follow up, since it requires changing those tests quite a bit.

Reply:
Hmm, I'd say if the tests rely on non-const values, then the tests need to be reworked somewhat (I assume they're just re-using existing candidates as opposed to creating new ones for the sake of simplicity). Either that or they're testing state-changes that aren't possible.

Agree that it's not worth refactoring the tests in this PR, but could you maybe organize the members and add a note to the definition of `NodeEvictionCandidate` to make the possible state changes clear? I *think* there are basically 3 types, though I may be oversimplifying:
- These never change after creation
- These may change once
- These may change at any time